### PR TITLE
Authorize call and app identifier changes for singleton service

### PR DESCRIPTION
**Pull request by NomadXD**
This PR includes,

- cache update operation by performing http calls to 3scale SM API per application. 
- delta store modified to work with app identifier.
- singleton service configuration splitted into several sub configs.
- report and auth call modified to work with app identifier types. (user_key, app_id , app_id+app_key)
- removed threescalers fork and added threescalers main

### CODE REVIEW

File: cache-filter/src/filter/http.rs [hunk 1]
```diff
@@ -301,7 +306,7 @@ impl Context for CacheFilter {
                             request_process_failure(self, self)
                         }
                     }
-                    Ok(Authorization::Denied(denied_auth)) => {
+                    Ok(Authorization::Error(denied_auth)) => {
                         info!("authorization was denied with code: {}", denied_auth.code());
                         request_process_failure(self, self);
                         return;
```
Comment (unleashed) [inline cache-filter/src/filter/http.rs:312]: This is not a denied authorization, but an error in the authorization call - ie. missing application identifier/user_key, missing service identifier or wrong service token, specified a missing metric, etc.

File: cache-filter/src/filter/http.rs [hunk 2]
```diff
@@ -292,7 +297,7 @@ impl Context for CacheFilter {
         match self.get_http_call_response_body(0, body_size) {
             Some(bytes) => {
                 match Authorization::from_str(std::str::from_utf8(&bytes).unwrap()) {
-                    Ok(Authorization::Ok(response)) => {
+                    Ok(Authorization::Status(response)) => {
                         if let Err(e) = self.handle_auth_response(&response) {
```
Comment (unleashed) [inline cache-filter/src/filter/http.rs:301]: Note that you are passing in the `AuthorizationStatus` and `handle_auth_response` just goes on to pick the usage reports, but does not actually check that the authorization is accepted/authorized or rejected/denied.

File: cache-filter/src/filter/http.rs (hunk 1 above)
Comment (NomadXD) [reply cache-filter/src/filter/http.rs:312]: ops. last moment change of the threescalers api. will fix that info log.

File: cache-filter/src/filter/http.rs (hunk 2 above)
Comment (NomadXD) [reply cache-filter/src/filter/http.rs:301]: @unleashed This is actually on the cache filter side. I didn't change anything related to the implementation. Just changed the Authorization to use the new API. eg: `Authorization::Ok`  to `Authorization::Status`. And yes I think `handle_auth_response` should take account status as well. We will get both 200 and 409 through that `Authorization::Ok`. `handle_auth_response` in singleton service takes status into account. https://github.com/NomadXD/gsoc-wasm-filters/blob/dev/auth-call/singleton-service/src/service/proxy.rs#L252
Will fix this in a separate PR. cc @rahulanand16nov

Comment (rahulanand16nov) [reply cache-filter/src/filter/http.rs:301]: @NomadXD Better do it in this PR as it's a small change - let me know if you need any help with it.

File: cache-filter/src/filter/http.rs (hunk 1 above)
Comment (rahulanand16nov) [reply cache-filter/src/filter/http.rs:312]: Also, change the variable name above it.

File: singleton-service/src/service/auth.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    service::Service,
+    transaction::Transaction,
+};
+
+/// Proxy level representation for authorization request.
+pub struct Auth {
+    pub service_id: String,
+    pub service_token: String,
+    pub app_id: AppIdentifier,
+}
```
Comment (rahulanand16nov) [inline singleton-service/src/service/auth.rs:18]: You are publishing and providing getters as well. Have either not both.

File: singleton-service/src/service/auth.rs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+            debug!("AppIdentifier AppId : {}", auth.app_id().as_string());
+            app = Application::from_app_id(auth.app_id().as_string())
+        }
+        AppIdentifier::AppId(_app_id, _app_key) => {
+            let app_key = auth.app_id().as_string();
+            debug!("AppIdentifier AppId+AppKey : {}", app_key);
+            let keys = app_key.split('_').collect::<Vec<_>>();
+            app = Application::from_app_id_and_key(keys[0], keys[1])
+        }
+    }
```
Comment (rahulanand16nov) [inline singleton-service/src/service/auth.rs:75]: Can you please shed some light on the following concerns:
- You've pattern-matched `user_key` but not used it instead provided app_id in place of it?
- Even in the second pattern? you have required fields but still reforming them using `as_string()`?
- Same goes for 3rd pattern.
Maybe I am getting this wrong?

File: singleton-service/src/service/report.rs [hunk 5]
```diff
@@ -64,10 +64,26 @@ pub fn build_report_request(report: &Report) -> Result<Request, anyhow::Error> {
...
+                debug!("AppIdentifier AppId : {}", app_id.as_string());
+                app = Application::from_app_id(app_id.as_string())
+            }
+            AppIdentifier::AppId(_app_id, _app_key) => {
+                let app_key = app_id.as_string();
+                debug!("AppIdentifier AppId+AppKey : {}", app_key);
+                let keys = app_key.split('_').collect::<Vec<_>>();
+                app = Application::from_app_id_and_key(keys[0], keys[1])
+            }
+        }
```
Comment (rahulanand16nov) [inline singleton-service/src/service/report.rs:84]: Duplicate code with the same concerns mentioned above.

**Review by rahulanand16nov — CHANGES_REQUESTED** [review]

File: cache-filter/src/filter/http.rs (hunk 1 above)
Comment (NomadXD) [reply cache-filter/src/filter/http.rs]: fixed

File: cache-filter/src/filter/http.rs (hunk 2 above)
Comment (NomadXD) [reply cache-filter/src/filter/http.rs]: fixed

File: singleton-service/src/service/auth.rs (hunk 3 above)
Comment (NomadXD) [reply singleton-service/src/service/auth.rs]: fixed

File: singleton-service/src/service/auth.rs (hunk 4 above)
Comment (NomadXD) [reply singleton-service/src/service/auth.rs]: sorry this was a temp thing i was trying and forgot to fix it before the PR. fixed

File: singleton-service/src/service/report.rs (hunk 5 above)
Comment (NomadXD) [reply singleton-service/src/service/report.rs]: fixed

**Review by rahulanand16nov — APPROVED** [review]
LGTM now :)

**Pull request merged**